### PR TITLE
Add `copy` kwarg to `asarray`

### PR DESCRIPTION
**Pull request by neosunhan**
## Description
Enables `copy` kwarg in `asarray`. The keyword defaults to `False` for backward compatibility.

Issue/s resolved: #990

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence
- [ ] All split configurations tested
- [ ] Multiple dtypes tested in relevant functions
- [ ] Documentation updated (if needed)
- [ ] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

skip ci

### CODE REVIEW

File: heat/core/factories.py [hunk 1]
```diff
@@ -434,6 +434,7 @@ def array(
 def asarray(
     obj: Iterable,
     dtype: Optional[Type[datatype]] = None,
+    copy: bool = False,
```
Comment (ClaudiaComito) [inline heat/core/factories.py:437]: According to the array-API standard, `asarray(copy=False)` should return `ValueError` if a copy turns out to be necessary. 
https://data-apis.org/array-api/latest/API_specification/generated/signatures.creation_functions.asarray.html

I propose to set this default to `None` (we won't copy unless necessary), which is basically the default numpy users expect. Then if the users specify `asarray(copy=False)` and copying is necessary, we can return a `ValueError`.

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks a lot @neosunhan ! I only have a comment about the array-API behaviour when `copy=False`, let me know what you think.

**Comment by ClaudiaComito** [comment]
@JuanPedroGHM have we got a fitting label for these changes re: auto-generated changelog? I was going to suggest `array-API`. Also, I'm thinking we need some kind of roadblock for when we're about to merge a PR that hasn't been labelled properly.

File: heat/core/factories.py (hunk 1 above)
Comment (neosunhan) [reply heat/core/factories.py]: I've changed the default and added a check in the event `copy=False`.

File: heat/core/factories.py [hunk 2]
```diff
@@ -482,7 +487,15 @@ def asarray(
...
     """
-    return array(obj, dtype=dtype, copy=False, order=order, is_split=is_split, device=device)
+    if isinstance(copy, bool) and not copy:
+        if not (
+            isinstance(obj, DNDarray)
+            and (dtype is None or dtype == obj.dtype)
+            and (is_split is None or is_split == obj.split)
+            and (device is None or device == obj.device)
+        ):
+            raise ValueError("copy is necessary")
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/factories.py:497]: I think the check fits better in the array function where the real logic is.

**Review by mtar — CHANGES_REQUESTED** [review]
Now that we have a CI running again, we can finally review some PRs. :smiley:

**Comment by ClaudiaComito** [comment]
Switched base branch from `main` to `release/1.2.x`  so we can release this sooner.

**Comment by ClaudiaComito** [comment]
Closing this as it is addressed in #1119

**Pull request closed without merging**